include/inviwo/core/properties/positionproperty.h

```diff
@@ -53,54 +58,83 @@ class IVW_CORE_API PositionProperty : public CompositeProperty {
 
     using CompositeProperty::set;  // Enable calling CompositeProperty::set(...) functions even
                                    // though overriding with set(const vec3& value)
     /**
-     * \brief Set coordinate in world space.
-     * @param worldSpacePos Position in world space.
+     * Set both position \p pos given in the coordinate space \p space and the coordinate space.
+     * @param pos    position in \p space coordinates
+     * @param space  reference coordinate system of \p pos
+     */
+    void set(const vec3& pos, CoordinateSpace space);
+
+    /**
+     * Update the position given \p pos in coordinate space \p sourceSpace. The position will be
+     * transformed from \p sourceSpace to the currently set reference coordinate space of the
+     * property.
+     * @param pos    position in \p space coordinates
+     * @param sourceSpace  coordinate space of \p pos
+     */
+    void updatePosition(const vec3& pos, CoordinateSpace sourceSpace);
+
+    [[nodiscard]] CoordinateSpace getCoordinateSpace() const;
+
+    /**
+     * Determines the direction between origin and position in the current coordinate space and
+     * transforms it to world coordinates. In case of world coordinates, this is equivalent to the
+     * position itself.
+     * @return directional vector in world coordinates
+     */
+    [[nodiscard]] vec3 getWorldSpaceDirection() const;
+
+    /**
+     * Set the property semantics of the position property to \p semantics
      */
-    void set(const vec3& worldSpacePos);
+    void setPositionSemantics(PropertySemantics semantics);
 
-    virtual void serialize(Serializer& s) const override;
     virtual void deserialize(Deserializer& d) override;
 
-    OptionPropertyInt
-        referenceFrame_;          //< The space in which the position is specified (world or view).
-    FloatVec3Property position_;  //< Position in specified space (world or view).
 private:
-    void referenceFrameChanged();
-    void positionChanged();
-    void cameraChanged();
+    [[nodiscard]] vec3 convertFromWorld(const vec3& pos, CoordinateSpace targetSpace) const;
+    [[nodiscard]] vec3 convertToWorld(const vec3& pos, CoordinateSpace sourceSpace) const;
+    void invalidateOutputProperties();
+    void referenceSpaceChanged();
+
+    CameraProperty* camera_;  //< Non-owning reference.
+    std::optional<CoordinateSpace> previousReferenceSpace_;
+
+    OptionProperty<CoordinateSpace> referenceSpace_;
+    BoolProperty transformPosition_;
+    FloatVec3Property position_;
+
+    CompositeProperty output_;
+    FloatVec3RefProperty worldPos_;
+    FloatVec3RefProperty viewPos_;
+    FloatVec3RefProperty clipPos_;
+    FloatVec3RefProperty screenPos_;
 
-    vec3 positionWorldSpace_;  //< Used for always keeping track of the current position in world
-                               // space.
-    CameraProperty* camera_;   //< Non-owning reference.
+    std::vector<Property*> outputProperties_;
```

Comment:
remove, use a std::tie props thing, or a function that returns an array of the 4 pointers...